docs/src/proposals/optimistic_confirmation.md

```diff
@@ -0,0 +1,237 @@
+being voted for.
+
+Given a vote `vote(X, S)`, let `S.last == vote.last` be the last slot in `S`.
+
+* `Claim`: if a validator submits `vote(X, S)`, the same validator
+should not have voted on a slot `s'` where  `X < s' < S.last` and `s'`
+is not a descendant of `X` and an ancestor of `S.last`. 
+
+* `Proof`: First note that the vote for `S` must have come after the vote
+for `S'` (due to lockouts, the higher vote must have after). Thus, the sequence 
+of votes must have been: `X ... S' ... S`. This means after the second vote the
+validator must have switched back to the original fork at some `X' > X`. This 
+means the vote for `S` should have used `X'` as the "reference" point, so the 
+vote should have been of the form `vote(X', S)`, not `vote(X, S)`
+(Recall the vote should contain the `latest` switch).
+
+To enforce this, we define the "Optimistic Slashing" slashing conditions. Given
+any two distinct votes `vote(X, S)`and `vote(X', S')` by the same validator,
+the votes must satisfy (namely the last two rules imply the ranges cannot
+overlap):
+
+* `X <= S.last`, `X' <= S'.last`
+* All `s` in `S.last` are ancestors/descendants of one another, 
+all `s'` in `S'.last` areancsestors/desccendants of one another,
+* 
+* `X == X'` implies `S` is parent of `S'` or `S'` is a parent of `S`
+* `X' > X` implies `X' > S.last` and `S'.last > S.last` 
+and for all `s` in `S`, `s + lockout(s) < X'`
+* `X > X'` implies `X > S'.last` and `S.last > S'.last` 
+and for all `s` in `S'`, `s + lockout(s) < X`
+
+Otherwise the validator is slashed. 
+
+`Range(vote)` - Given a vote `v = vote(X, S)`, define `Range(v)` to be the range
+ of slots `[X, S.last]`.
+
+`SP(old_vote, new_vote)` - This is the "Switching Proof" for `old_vote`, the
+validator's latest vote. Such a proof is necessary for a validator to vote 
+on any slot `new_vote.last` that is not a descendant of `old_vote.last`. Note 
+switching must still respect lockouts.
+
+A switching proof shows that `> 1/3` of the network is locked out at slot
+`old_vote.last`.
+
+The proof is a list of elements `(validator_id, validator_vote(X, S))`, where:
+
+1. The sum of the stakes of all the validator id's `> 1/3`
+
+2. For each `(validator_id, validator_vote(X, S))`, there exists some slot `s`
+in `S` where: 
+    * a.`s` is not a common ancestor of both `validator_vote.last` and
+    `old_vote.last` and `new_vote.last`.
+    * b. `s` is not a descendant of `validator_vote.last`.
+    * c. `s + s.lockout() >= old_vote.last` (implies valdator is still locked
+    out on slot `s` at slot `old_vote.last`).
+
+Switching forks without a valid switching proof is slashable.
+
+## Definitions:
+
+Optimistic Confirmation - A block `B` is then said to have achieved
+"optimistic confirmation" if `>2/3` of stake have voted with votes `v`
+where `Range(v)` for each such `v` includes `B.slot`.
+
+Finalized - A block `B` is said to be finalized if at least one 
+correct validator has rooted `B` or a descedant of `B`.
```

Comment:
descendant

Reply:
what is a "correct validator"?

Reply:
It's a validator that is following the protocol as written, not acting in a byzantine way.